tests/bluetooth/bsim_bt/bsim_test_gatt_caching/src/gatt_client_test.c

```diff
@@ -0,0 +1,363 @@
+{
+	static struct bt_gatt_discover_params discover_params;
+	int err;
+
+	printk("Discovering services and characteristics\n");
+
+	discover_params.uuid = uuid;
+	discover_params.func = discover_func;
+	discover_params.start_handle = BT_ATT_FIRST_ATTRIBUTE_HANDLE;
+	discover_params.end_handle = BT_ATT_LAST_ATTRIBUTE_HANDLE;
+	discover_params.type = type;
+
+	UNSET_FLAG(flag_discover_complete);
+
+	err = bt_gatt_discover(g_conn, &discover_params);
+	if (err != 0) {
+		FAIL("Discover failed(err %d)\n", err);
+	}
+
+	WAIT_FOR_FLAG(flag_discover_complete);
+	printk("Discover complete\n");
+}
+
+static uint8_t gatt_read_cb(struct bt_conn *conn, uint8_t err, struct bt_gatt_read_params *params,
+			    const void *data, uint16_t length);
+
+static struct {
+	struct bt_gatt_read_params params;
+	uint8_t expected_status;
+} reads[] = {
+	{
+		 /* Read on channel 1 */
+		.params = {
+			.func = gatt_read_cb,
+			.handle_count = 1,
+			.single = {
+				.handle = 0, /* Will be set later */
+				.offset = 0,
+			},
+		},
+		.expected_status = BT_ATT_ERR_DB_OUT_OF_SYNC,
+	},
+	{
+		 /* Read on channel 2 */
+		.params = {
+			.func = gatt_read_cb,
+			.handle_count = 1,
+			.single = {
+				.handle = 0, /* Will be set later */
+				.offset = 0,
+			},
+		},
+		.expected_status = BT_ATT_ERR_DB_OUT_OF_SYNC,
+	},
+	{
+		 /* DB hash read */
+		.params = {
+			.func = gatt_read_cb,
+			.handle_count = 0,
+			.by_uuid = {
+				.start_handle = BT_ATT_FIRST_ATTRIBUTE_HANDLE,
+				.end_handle = BT_ATT_LAST_ATTRIBUTE_HANDLE,
+				.uuid = BT_UUID_GATT_DB_HASH,
+			},
+		},
+		.expected_status = BT_ATT_ERR_SUCCESS,
+	},
+	{
+		 /* Read on channel 1 */
+		.params = {
+			.func = gatt_read_cb,
+			.handle_count = 1,
+			.single = {
+				.handle = 0, /* Will be set later */
+				.offset = 0,
+			},
+		},
+		.expected_status = BT_ATT_ERR_SUCCESS,
+	},
+};
+
+static volatile uint8_t read_index;
+
+static uint8_t gatt_read_cb(struct bt_conn *conn, uint8_t err, struct bt_gatt_read_params *params,
+			    const void *data, uint16_t length)
+{
+	printk("GATT read cb: err 0x%02X\n", err);
+	if (err != reads[read_index].expected_status) {
+		FAIL("Unexpected status from read: 0x%02X\n", err);
+	}
+
+	if (params != &reads[read_index].params) {
+		FAIL("Wrong read\n");
+	}
+
+	(void)memset(params, 0, sizeof(*params));
+
+	read_index++;
+
+	return 0;
+}
+
+static void gatt_read(struct bt_gatt_read_params *read_params)
+{
+	int err;
+
+	printk("Reading\n");
+
+	err = bt_gatt_read(g_conn, read_params);
+	if (err != 0) {
+		FAIL("bt_gatt_read failed: %d\n", err);
+	}
+}
+
+static void write_cb(struct bt_conn *conn, uint8_t err, struct bt_gatt_write_params *params)
+{
+	if (err != BT_ATT_ERR_SUCCESS) {
+		FAIL("Write failed: 0x%02X\n", err);
+	}
+
+	SET_FLAG(flag_write_complete);
+}
+
+static void enable_robust_caching(void)
+{
+	/* Client Supported Features Characteristic Value
+	 * Bit 0: Robust Caching
+	 * Bit 1: EATT
+	 */
+	static const uint8_t csf[] = { BIT(0) | BIT(1) };
```

Comment:
They are not public, defined in `gatt.c`.

I think we should consider automatically writing the Client Supported Features Characteristic so that the application doesn't have to do this, but that is out of scope here.

Reply:
Indeed. `CF_EATT` does not seem to be used at all either. 

Will you create an issue for this?

Reply:
Can create an issue.

For LE, reading the EATT bit is not needed before attempting to connect an EATT channel so it it mostly useless. I think it is required for BR/EDR though.

Reply:
Maybe it only holds for the Server Supported Feature Characteristic?

>Vol 3, Part G: GATT, 6.2.1 Connection Establishment:
>
>Note: Unlike BR/EDR, it is not necessary to check the Server Supported Features characteristic before attempting to establish an Enhanced ATT bearer.

Reply:
> Will you create an issue for this?

https://github.com/zephyrproject-rtos/zephyr/issues/44587